src/relay/dnsrelay/mod.rs

```diff
@@ -0,0 +1,260 @@
+
+    query.set_query_type(qtype);
+    query.set_name(qname.clone());
+
+    message.set_id(6666);
+    message.set_recursion_desired(true);
+    message.add_query(query);
+
+    let req_buffer = message.to_vec()?;
+    socket.send_to(&req_buffer, server).await?;
+
+    let mut res_buffer = vec![0; 512];
+    socket.recv_from(&mut res_buffer).await?;
+
+    Ok(Message::from_vec(&mut res_buffer)?)
+}
+
+async fn socks5_lookup(qname: &Name, qtype: RecordType, socks5: SocketAddr, ns: SocketAddr) -> Result<Message> {
+    let mut stream = TcpStream::connect(socks5).await?;
+
+    // 1. Handshake
+    let hs = HandshakeRequest::new(vec![SOCKS5_AUTH_METHOD_NONE]);
+    hs.write_to(&mut stream).await?;
+    stream.flush().await?;
+
+    let hsp = HandshakeResponse::read_from(&mut stream).await?;
+    assert_eq!(hsp.chosen_method, SOCKS5_AUTH_METHOD_NONE);
+
+    // 2. Send request header
+    let addr = Address::SocketAddress(ns);
+    let h = TcpRequestHeader::new(Command::TcpConnect, addr);
+    h.write_to(&mut stream).await?;
+    stream.flush().await?;
+
+    let hp = TcpResponseHeader::read_from(&mut stream).await?;
+    match hp.reply {
+        Reply::Succeeded => (),
+        r => {
+            let err = io::Error::new(io::ErrorKind::Other, format!("{}", r));
+            return Err(err);
+        }
+    }
+
+    let mut message = Message::new();
+    let mut query = Query::new();
+
+    query.set_query_type(qtype);
+    query.set_name(qname.clone());
+
+    message.set_id(6666);
+    message.set_recursion_desired(true);
+    message.add_query(query);
+
+    let req_buffer = message.to_vec()?;
+    let size = req_buffer.len();
+    let mut size_buffer: [u8; 2] = [((size >> 8) & 0xFF) as u8, ((size >> 0) & 0xFF) as u8];
+    let mut send_buffer: [u8; 512 + 2] = [0; 512 + 2];
+    send_buffer[..2].copy_from_slice(&size_buffer[..2]);
+    send_buffer[2..size + 2].copy_from_slice(&req_buffer[0..size]);
+    stream.write_all(&send_buffer[0..size + 2]).await?;
+
+    stream.read_exact(&mut size_buffer[0..2]).await?;
+
+    let mut res_buffer = vec![0; 512];
+    let size = ((size_buffer[0] as usize) << 8) + (size_buffer[1] as usize);
+    stream.read_exact(&mut res_buffer[0..size]).await?;
+
+    Ok(Message::from_vec(&mut res_buffer)?)
+}
+
+async fn acl_lookup(
+    context: &SharedContext,
+    local: SocketAddr,
+    remote: SocketAddr,
+    socks5: SocketAddr,
+    qname: &Name,
+    qtype: RecordType,
+) -> Result<Message> {
+    // Start querying name servers
+    debug!(
+        "attempting lookup of {:?} {} with ns {} and {}",
+        qtype, qname, local, remote
+    );
+
+    let timeout = Some(Duration::new(5, 0));
+
+    let local_response = try_timeout(udp_lookup(qname, qtype.clone(), local), timeout)
+        .await
+        .unwrap_or(Message::new());
+    let remote_response = try_timeout(socks5_lookup(qname, qtype.clone(), socks5, remote), timeout)
+        .await
+        .unwrap_or(Message::new());
```

Comment:
We should send requests first, then await `local_response` and if we know we are going to bypass, we can cancel `remote_response`.

Reply:
Of course, this is lower priority...